### ISSUE TITLE
Enters infinite loop when flow-type definition references itself

### ISSUE THREAD

**Issue by aray12** [yannickcr/eslint-plugin-react#1032]
I found that if you try to create prop-types using flowtype definitions and accidentally reference yourself within it you enter into a infinite loop and run out of memory

e.g.
```javascript
type Props = {
  someProp: string,
  anotherProp: Props,
};
```

You get:
```
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at buildTypeAnnotationDeclarationTypes (/node_modules/eslint-plugin-react/lib/rules/prop-types.js:403:35)
```

Obviously the application code itself is broken, but I think this should fail gracefully. 

If this is a problem I should be reporting directly to eslint, please let me know

**Comment by ljharb** [comment]
I think this is a bug; the recursion should keep a `Set` of seen values, to handle cycles.

**Comment by minedeljkovic** [comment]
> Obviously the application code itself is broken, but I think this should fail gracefully

There is a valid need for recursive types even in prop-types, it is not always result of a mistake. We have a lot in our app, so we had to temporary remove `react/prop-types` rule.

Older dupe of this is #913

**Comment by ljharb** [comment]
Closing as a dupe.
